### PR TITLE
Fix race in Slintpad causing preview to panic (esp. in FF)

### PR DESCRIPTION
**Pull request by tronical**
With the following snippet there was a race:

```
if (instance === null) {
    instance = await create();
    await instance.show();
} else {
    instance.create_with_existing_window(...);
}
```

When awaiting for the promise that show() returns to settle, we would re-enter this function (due to a broken recursion guard, that's now removed). When re-entering, create_with_existing_window() is called, which - in order to access the slint::Window to re-use - must create the window adapter. Since that didn't happen from within an event loop invocation, winit would panic.

Consequently, this patch set also makes create_with_existing_window() return a promise, as well as hide() (for the sake of it).

The broken recursion guard is easy to fix (with a .finally() handler setting the boolean back), but that would mean we end up not always rendering the very latest sources. Plus, the same issue exists in wasm_preview.ts.

Now, in case of subsequent re-entrancy in this code, we end up creating a bunch of create_with_exiting_window() promises, which will eventually settle and then we go back to the original show() call that will show the canvas.

### CODE REVIEW

**Comment by ogoffart** [comment]
don't you still get a recursion if the function is called while create is awaiting?

I'm thinking you need to do something like this:
(not sure if this is valid to await several time on a promise)

```js
let i = await instance;
if (i === null) {
    instance = create();
    await (await instance).show();
} else {
    await i.create_with_existing_window(...);
}
```

File: tools/slintpad/src/lsp.ts [hunk 1]
```diff
@@ -213,8 +213,15 @@ class PreviewerBackend {
                     });
                 },
             );
-            this.#instance.set_design_mode(this.#picker_mode);
+            instance.set_design_mode(this.#picker_mode);
+        })
+    }
+
+    private async with_instance<R>(callback: (instance: slint_preview.WrappedInstance) => R): Promise<R | null> {
```
Comment (ogoffart) [inline tools/slintpad/src/lsp.ts]: the warning looks indeed like something's fishy.

Comment (tronical) [reply tools/slintpad/src/lsp.ts]: The warning is bogus AFAICS. Watch how eslint complains in this commit:

https://github.com/slint-ui/slint/pull/2879/commits/2681768d10e45d04987e58b116020a3f5eee82e6

claiming that `instance` is unused here:

```typescript
type InstanceCallback<R> = (_instance: slint_preview.WrappedInstance) => R;
```

and https://github.com/slint-ui/slint/pull/2879/commits/309ff0f3aec973f474c4eada3d13558168edc87a will silence it again.

**Review by ogoffart — APPROVED** [review]
And the docs don't need to be adjusted because everything is already async there?

**Comment by tronical** [comment]
Good point, I'll fix up the docs.

**Pull request merged**